cmd/cosign/cli/verify/verify_blob.go

```diff
@@ -161,44 +162,74 @@ func VerifyBlobCmd(ctx context.Context, ko sign.KeyOpts, certRef, certEmail, cer
 		if len(uuids) == 0 {
 			return errors.New("could not find a tlog entry for provided blob")
 		}
+		return verifySigByUUID(ctx, ko, rClient, certEmail, certOidcIssuer, sig, b64sig, uuids, blobBytes)
+	}
+
+	// Use the DSSE verifier if the payload is a DSSE with the In-Toto format.
+	if isIntotoDSSE(blobBytes) {
+		verifier = dsse.WrapVerifier(verifier)
+	}
+
+	// verify the signature
+	if err := verifier.VerifySignature(bytes.NewReader([]byte(sig)), bytes.NewReader(blobBytes)); err != nil {
+		return err
+	}
+
+	// verify the rekor entry
+	if err := verifyRekorEntry(ctx, ko, verifier, cert, b64sig, blobBytes); err != nil {
+		return err
+	}
+
+	fmt.Fprintln(os.Stderr, "Verified OK")
+	return nil
+}
 
-		tlogEntry, err := cosign.GetTlogEntry(ctx, rClient, uuids[0])
+func verifySigByUUID(ctx context.Context, ko sign.KeyOpts, rClient *client.Rekor, certEmail, certOidcIssuer, sig, b64sig string, uuids []string, blobBytes []byte) error {
+	var validSigExists bool
+	for _, u := range uuids {
+		tlogEntry, err := cosign.GetTlogEntry(ctx, rClient, u)
 		if err != nil {
-			return err
+			continue
 		}
 
 		certs, err := extractCerts(tlogEntry)
 		if err != nil {
-			return err
+			continue
 		}
 
 		co := &cosign.CheckOpts{
 			RootCerts:      fulcio.GetRoots(),
 			CertEmail:      certEmail,
 			CertOidcIssuer: certOidcIssuer,
 		}
-		cert = certs[0]
-		verifier, err = cosign.ValidateAndUnpackCert(cert, co)
+		cert := certs[0]
+		verifier, err := cosign.ValidateAndUnpackCert(cert, co)
 		if err != nil {
-			return err
+			continue
+		}
+		// Use the DSSE verifier if the payload is a DSSE with the In-Toto format.
+		if isIntotoDSSE(blobBytes) {
+			verifier = dsse.WrapVerifier(verifier)
+		}
+		// verify the signature
+		if err := verifier.VerifySignature(bytes.NewReader([]byte(sig)), bytes.NewReader(blobBytes)); err != nil {
+			continue
 		}
-	}
 
-	// Use the DSSE verifier if the payload is a DSSE with the In-Toto format.
-	if isIntotoDSSE(blobBytes) {
-		verifier = dsse.WrapVerifier(verifier)
+		// verify the rekor entry
+		if err := verifyRekorEntry(ctx, ko, verifier, cert, b64sig, blobBytes); err != nil {
```

Comment:
FYI inside this we make another extraneous call to Rekor to get the TLOG entry UUID (https://github.com/sigstore/cosign/blob/e2017af9d4ed79c3e15d944eb19a818905af47c2/cmd/cosign/cli/verify/verify_blob.go#L326) and perform tlog verification (check inclusion proof etc) inside this. I had this refactored/separated out https://github.com/sigstore/cosign/pull/1662 so I'll rebase and make it even more clean here!

Reply:
thanks @asraa !